nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearch.java

```diff
@@ -0,0 +1,193 @@
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(
+                    AttributeExpression.ResultType.STRING, true))
+            .build();
+
+    public static final PropertyDescriptor TYPE = new PropertyDescriptor.Builder()
+            .name("Type")
+            .description("The type of this document (used by Elasticsearch for indexing and searching)")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(
+                    AttributeExpression.ResultType.STRING, true))
+            .build();
+
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        relationships.add(REL_RETRY);
+        return Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(CLUSTER_NAME);
+        descriptors.add(HOSTS);
+        descriptors.add(PROP_SSL_CONTEXT_SERVICE);
+        descriptors.add(PROP_SHIELD_LOCATION);
+        descriptors.add(PING_TIMEOUT);
+        descriptors.add(SAMPLER_INTERVAL);
+        descriptors.add(DOC_ID);
+        descriptors.add(INDEX);
+        descriptors.add(TYPE);
+
+        return Collections.unmodifiableList(descriptors);
+    }
+
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        // Create the client if one does not already exist
+        createClient(context);
+
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions(flowFile).getValue();
+        final String docId = context.getProperty(DOC_ID).evaluateAttributeExpressions(flowFile).getValue();
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(flowFile).getValue();
+
+        final ProcessorLog logger = getLogger();
+        try {
+
+            logger.debug("Fetching {}/{}/{} from Elasticsearch", new Object[]{index, docType, docId});
+            final GetResponse getResponse =
+                    esClient.get().prepareGet(index, docType, docId).execute().actionGet();
+
+            if (getResponse == null || !getResponse.isExists()) {
+                logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                        new Object[]{index, docType, docId});
+                session.transfer(flowFile, REL_RETRY);
+                context.yield();
```

Comment:
Will change to yield. We would want to retry on not found in case NiFi gets the document ID but the ES index has not yet been updated to reflect the addition of the document (i.e. race condition)